gax/src/main/java/com/google/api/gax/rpc/mtls/MtlsProvider.java

```diff
@@ -0,0 +1,191 @@
+        DEFAULT_CONTEXT_AWARE_METADATA_PATH);
+  }
+
+  /**
+   * Returns if mutual TLS client certificate should be used. If the value is true, the key store
+   * from {@link #getKeyStore()} will be used to configure mutual TLS transport.
+   */
+  public boolean useMtlsClientCertificate() {
+    String useClientCertificate = envProvider.getenv("GOOGLE_API_USE_CLIENT_CERTIFICATE");
+    return "true".equals(useClientCertificate);
+  }
+
+  /**
+   * Returns AUTO, NEVER or ALWAYS. NEVER means always use regular endpoint; ALWAYS means always use
+   * mTLS endpoint; AUTO means auto switch to mTLS endpoint if client certificate exists and should
+   * be used.
+   */
+  public UseMtlsEndpoint useMtlsEndpoint() {
+    String useMtlsEndpoint = envProvider.getenv("GOOGLE_API_USE_MTLS_ENDPOINT");
+    if ("never".equals(useMtlsEndpoint)) {
+      return UseMtlsEndpoint.NEVER;
+    } else if ("always".equals(useMtlsEndpoint)) {
+      return UseMtlsEndpoint.ALWAYS;
+    }
+    return UseMtlsEndpoint.AUTO;
+  }
+
+  /** The mutual TLS key store created with the default client certificate on device. */
+  public KeyStore getKeyStore() throws IOException {
+    try (InputStream stream = new FileInputStream(metadataPath)) {
+      return getKeyStore(stream, processProvider);
+    } catch (FileNotFoundException exception) {
+      // file doesn't exist
+      return null;
+    }
+  }
+
+  @VisibleForTesting
+  static KeyStore getKeyStore(InputStream metadata, ProcessProvider processProvider)
+      throws IOException {
+    try {
+      Process process = processProvider.createProcess(metadata);
+
+      // Run the command and timeout after 1000 milliseconds.
+      int exitCode = runCertificateProviderCommand(process, 1000);
+      if (exitCode != 0) {
+        throw new IOException("Cert provider command failed with exit code: " + exitCode);
+      }
+
+      // Create mTLS key store with the input certificates from shell command.
+      return SecurityUtils.createMtlsKeyStore(process.getInputStream());
+    } catch (InterruptedException e) {
+      throw new IOException("Interrupted executing certificate provider command", e);
+    } catch (GeneralSecurityException e) {
+      throw new IOException("Failed to get key store", e);
+    }
+  }
+
+  @VisibleForTesting
+  static ImmutableList<String> extractCertificateProviderCommand(InputStream contextAwareMetadata)
+      throws IOException {
+    JsonParser parser = new GsonFactory().createJsonParser(contextAwareMetadata);
+    ContextAwareMetadataJson json = parser.parse(ContextAwareMetadataJson.class);
+    return json.getCommands();
+  }
+
+  @VisibleForTesting
+  static int runCertificateProviderCommand(Process commandProcess, long timeoutMilliseconds)
+      throws IOException, InterruptedException {
+    long startTime = System.currentTimeMillis();
+    long remainTime = timeoutMilliseconds;
+
+    while (remainTime > 0) {
+      Thread.sleep(Math.min(remainTime + 1, 100));
```

Comment:
Curious, what is the motivation behind this particular sleep duration? Should we add a brief comment?

Reply:
this information should be added to the comment

Reply:
done